theano/contrib/mkl/mkl_ndarray.h

```diff
@@ -54,16 +59,18 @@ char* MKL_TYPE[] = {"", "", "", "int16", "", "int32", "", "int64",
 typedef struct __MKLNdarray__{
 
     PyObject_HEAD
-    PyObject * base;
+    PyObject * base;        // reference for private_layout/data/workspace
 
     /* Type-specific fields go here. */
-    int         nd;                             // the number of dimensions of the tensor, maximum is 16 (MAX_NDIM).
-    int         dtype;                          // an integer type number is given here.
-    size_t      data_size;                      // the number of bytes allocated for mkl_data
-    size_t      user_structure[2 * MAX_NDIM];   // user layout: [size0, size1, ..., stride0, stride1, ..., 0, 0].
-    dnnLayout_t private_layout;                 // layout instance create by MKL APIs
-    void*       private_data;                   // data buffer
-    void*       private_workspace;              // computation workspace for forward and backward
+    int         nd;                                 // the number of dimensions of the tensor, maximum is 16 (MNDA_MAX_NDIM).
+    int         dtype;                              // an integer type number is given here.
+    size_t      data_size;                          // the number of bytes allocated for private_data
+    size_t      workspace_size;                     // the number of bytes allocated for private_workspace
+    size_t      user_structure[2 * MNDA_MAX_NDIM];  // user layout: [size0, size1, ..., stride0, stride1, ..., 0, 0].
+    dnnLayout_t private_layout;                     // layout instance create by MKL APIs
+    void*       private_data;                       // data buffer
+    dnnLayout_t private_layout_ws;                  // layout for workspace
```

Comment:
Can you confirmt this extra layout is needed as sometimes the data layout and the workspace layout are different?

Currently, theano handle different workspace, but the way you do it is hidded from Theano. So it is fine. Is it possible that different node in the graph share the same workspace to lower the memory usage? I don't think it is mandatory, but if it is possible/already done, it would be great.